exonum/src/events/noise/mod.rs

```diff
@@ -143,22 +143,29 @@ impl Handshake for NoiseHandshake {
     }
 }
 
-fn read<S: AsyncRead + 'static>(sock: S) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
-    let buf = vec![0u8; HANDSHAKE_HEADER_LENGTH];
-    // First byte of handshake message is payload length, remaining bytes [1; len] is
-    // the handshake payload. Therefore, we need to read first byte and after that
-    // remaining payload.
-    read_exact(sock, buf).and_then(|(stream, msg)| read_exact(stream, vec![0u8; msg[0] as usize]))
-}
+pub struct HandshakeRawMessage(Vec<u8>);
+
+impl HandshakeRawMessage {
+    pub fn read<S: AsyncRead + 'static>(
+        sock: S,
+    ) -> impl Future<Item = (S, Self), Error = io::Error> {
+        let buf = vec![0u8; HANDSHAKE_HEADER_LENGTH];
+        // First byte of handshake message is payload length, remaining bytes [1; len] is
+        // the handshake payload. Therefore, we need to read first byte and after that
+        // remaining payload.
+        read_exact(sock, buf)
+            .and_then(|(stream, msg)| read_exact(stream, vec![0u8; msg[0] as usize]))
+            .and_then(|(stream, msg)| Ok((stream, HandshakeRawMessage(msg))))
+    }
 
-fn write<S: AsyncWrite + 'static>(
-    sock: S,
-    buf: &[u8],
-    len: usize,
-) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
-    debug_assert!(len < NOISE_MAX_HANDSHAKE_MESSAGE_LENGTH);
+    pub fn write<S: AsyncWrite + 'static>(
+        self,
+        sock: S,
+    ) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
+        let len = self.0.len();
+        debug_assert!(len < NOISE_MAX_HANDSHAKE_MESSAGE_LENGTH);
 
-    let mut message = vec![len as u8; HANDSHAKE_HEADER_LENGTH];
-    message.extend_from_slice(&buf[0..len]);
-    write_all(sock, message)
+        write_all(sock, vec![len as u8; HANDSHAKE_HEADER_LENGTH])
```

Comment:
Kind reminder: the code in `read` implicitly fixes `HANDSHAKE_HEADER_LENGTH` as 1, so it could make sense to get rid of the constant and simplify the code accordingly.

Reply:
We have plans to use handshake messages with payload, therefore header length will be increased to 4(or maybe 8) bytes (https://github.com/blackbeam/exonum-1/blob/032cf1afc69b4bc77b381821db9c1b5cb359d73f/exonum/src/events/noise/mod.rs#L71). And so I decided to leave it.

Reply:
But the code doesn't make sense with `HANDSHAKE_HEADER_LENGTH != 1`. I'm not sure it makes sense to generalize via `HANDSHAKE_HEADER_LENGTH` in general; you need to convert `len` to some type an write it (e.g., in little-endian).